### PR TITLE
persistent js runtime

### PR DESCRIPTION
**Pull request by letmaik**
very experimental, work-in-progress

TODO:
- re-create runtime whenever js app is redeployed (to avoid using old cached modules)
- quickjs [does not support](https://www.freelists.org/post/quickjs-devel/stackoverflow,2) being used from different threads -> evaluate against ccf's upcoming multi-threading support
  - maintain separate runtimes per thread
- evaluate perf improvements

### CODE REVIEW

File: src/apps/js_generic/js_generic.cpp [hunk 1]
```diff
@@ -1004,6 +967,47 @@ namespace ccfapp
...
+            fmt::format(
+              "No handler script found for method '{}'.", verb_prefixed));
+          return;
+        }
+      }
+
+      // Ideally this should happen in the constructor but that leads
+      // to issues with quick's stack overflow checks.
+      // This is only safe as long as this instance of JSHandlers
+      // is only called from a single persistent thread.
```
Comment (achamayou) [inline src/apps/js_generic/js_generic.cpp:1004]: That's fine, the runtime can be created once per thread and owned by a unique pointer in thread-local storage for example.

**Comment by letmaik** [comment]
It looks like quickjs' stack overflow test is not compatible with storing a runtime instance and then later entering the same path again if the stack size is not constant. Since CCF doesn't guarantee a constant stack size when reaching an endpoint handler the only solution would be to disable the stack overflow check, although this would mean infinite loops in JS would crash the node instead of being detected. Since C++ apps also do not have this protection it may make sense to disable the check.

File: src/apps/js_generic/js_generic.cpp (hunk 1 above)
Comment (letmaik) [reply src/apps/js_generic/js_generic.cpp]: I made it thread-local, could you have a quick look if this is fine?  I also found a way to deal with the stack overflow check, see comments in the code.

**Comment by letmaik** [comment]
I think I'm going to abandon this. I did some benchmarks locally with the forum app and could not see improvements.

**Pull request closed without merging**